specs/jsonschema-core.md

```diff
@@ -2029,32 +2029,129 @@ SHOULD use the terms defined by this document to do so.
+(`https:` to fetch schemas over HTTP, or `file:` to read schemas from disk),
+this behavior is discouraged unless it's intrinsic to the use case, such as with
+JSON Hyper-Schema.
+
+Resolving schemas dynamically introduces several security concerns, each of
+which can be mitigated by limiting or controlling resolution behavior. A tightly
+scoped schema resolution policy significantly reduces the attack surface,
+especially when validating untrusted data.
+
+Implementations are advised to disable dynamic retrieval by default and limit
+external schema resolution to the local registry unless dynamic retrieval is
+explicitly enabled. If enabled, they should consider limiting the number of
+dynamic retrievals a validation can perform and defining timeouts on dynamic
+retrievals to reduce the risk of resource exhaustion.
+
+#### HTTP(S) Specific Threats
+
+Allowing schema references to resolve over HTTP or HTTPS introduces several
+threats:
+
+* **Denial of Service (DoS)**: Validation may hang or become slow if a
+  referenced schema URL is slow to respond or never returns.
+* **Server-Side Request Forgery (SSRF)**: Malicious schemas can reference
+  internal-only services using hostnames like localhost or private IPs.
+  Implementations are advised to restrict HTTP schema retrieval to a
+  configurable allowlist of trusted domains.
+* **Lack of Integrity Guarantees**: Retrieved schemas may be altered in transit
+  or change between validations. If network retrieval is allowed,
+  implementations are advised to only allow retrieval over HTTPS unless
+  specifically configured to allow unsecured transport.
+
+#### File System Specific Threats
+
+Allowing resolution from the local filesystem (`file:` URIs) raises different
+issues:
+
+* **Information Disclosure**: Malicious schemas may access sensitive files on
+  the system. Implementations should consider restricting filesystem access to
+  a specific schema directory tree.
+* **Cross-Context Access**: A schema fetched from HTTP may try to reference a
+  schema on the filesystem. Implementations are advised to allow resolving
+  `file:` references only when the referencing schema was itself loaded from the
+  file system, similar to same-origin policies in web browsers.
+* **Exposing Internal Paths**: Schemas that use `file:` URIs may reveal
+  host-specific filesystem details in two ways: through the `$id` itself or
+  through schema locations in validation output. Implementations are advised to
+  reject `$id` values that use the `file:` scheme. If `file:` URIs are permitted
+  internally, implementations are advised to sanitize them (for example, by
+  converting them to relative URIs) to avoid exposing host filesystem structure
+  to users.
+
+### Vocabulary-Specific Risks
+
+Third-party JSON Schema vocabularies may introduce additional risks.
+Implementers are advised to consult the specifications of any extensions they
+support and take into account their security considerations as well.
```

Comment:
We should update this to use "extension" language since we've removed "vocabulary".